hs-bindgen/src-internal/HsBindgen/C/Fold/Type.hs

```diff
@@ -575,6 +483,84 @@ updateDeclAddAliases aliases = \case
 omapInsertBack :: Ord k => k -> v -> OMap.OMap k v -> OMap.OMap k v
 omapInsertBack k v m = m OMap.>| (k, v)
 
+{-------------------------------------------------------------------------------
+  Internal auxiliary: classify type declarations
+-------------------------------------------------------------------------------}
+
+data TypeDeclFlavour =
+    -- | Type declaration for which we have external bindings
+    TypeDeclExternal ExtIdentifier
+
+    -- | Opaque type declaration
+  | TypeDeclOpaque CName
+
+    -- | All other cases
+  | TypeDeclRegular
+
+-- | Anonymous top-level declaration
+--
+-- Since (even in C) these are unusable, we replace them by essentially an
+-- arbitrary type.
+--
+-- TODO: Should we warn when we see these?
+data AnonTopLevel = AnonTopLevel Type
+
+-- | Classify type declaration
+classifyTypeDecl ::
+     DeclPathCtxt
+  -> ExtBindings
+  -> (CXType, CXCursor, Either CInt CXCursorKind, SingleLoc)
```

Comment:
why a tuple?

Reply:
I use this a way to signal that these are all closely related.